Rename request body min rate APIs (#1901)

cesarbs opened the pull request:
#1901

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -253,13 +253,52 @@ public TimeSpan RequestHeadersTimeout
-        public MinimumDataRate RequestBodyMinimumDataRate { get; set; } = new MinimumDataRate(rate: 1, gracePeriod: TimeSpan.FromSeconds(5));
+        public MinimumDataRate MinRequestBodyDataRate { get; } = new MinimumDataRate { BytesPerSecond = 1, GracePeriod = TimeSpan.FromSeconds(5) };
+
+        public class MinimumDataRate
```
I don't like having a public nested class.

Tratcher replied:
The feature uses two fields but the options use a custom type? Why is this type needed for options?

cesarbs replied:
@Tratcher `MinRequestBodyDataRate.BytesPerSecond` matches how the feature is used better than `MinRequestBodyDataRateBytesPerSecond`. It's just aesthetics though.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -253,13 +253,52 @@ public TimeSpan RequestHeadersTimeout
+                get => _gracePeriod;
+                set
+                {
+                    if (value < TimeSpan.Zero)
```
I wonder if we should throw for TimeSpan.Zero too, since that's just asking for problems.

cesarbs replied:
It's still a legal value. Means the rate will be enforced right away. It's not recommended, but there's no technical reason to disallow it.

halter73 replied:
The technical reason to disallow it is because setting it to zero creates a race that could cause any connection to be killed no matter the actual upload rate

cesarbs replied:
Then we should require it to be larger than 1 second.

cesarbs replied:
Or, more precisely, larger than `Heartbeat.Interval`, which right now is one second.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.FeatureCollection.cs:
```diff
@@ -229,10 +229,32 @@ string IHttpRequestIdentifierFeature.TraceIdentifier
+            get => MinRequestBodyDataRate.GracePeriod;
+            set
+            {
+                if (value < TimeSpan.Zero)
```
Is there a way to share the validation logic between this and the MinimumDataRate class?

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpMinRequestBodyDataRateFeature.cs:
```diff
@@ -1,18 +1,25 @@
+        double BytesPerSecond { get; set; }
+
+        /// <summary>
+        /// The amount of time to delay enforcement of the minimum data rate.
```
Starting from when? After the request arrives? The first read?

cesarbs replied:
Will add ", starting at the time data is first read or written." Written as well since this same type will (likely) be used for the response rate.

cesarbs commented:
- Reverted 5c63a585a66368c2c232602c6147d8b26e347786 to stop the bikeshedding
- Moved `MinimumDataRate` to the base Core namespace so people don't need a weird `using` when setting it in `KestrelServerLimits`
- Allow 0 bytes/second
- Require the grace period to be larger than `Heartbeat.Interval`

cesarbs commented:
📢

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpMinRequestBodyDataRateFeature.cs:
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Features
     /// Represents a minimum data rate for the request body of an HTTP request.
     /// </summary>
-    public interface IHttpRequestBodyMinimumDataRateFeature
+    public interface IHttpMinRequestBodyDataRateFeature
```
I'm not sure we should be abbreviating Minimum to Min

halter73 replied:
After thinking about this for a little bit, I think we should change all usages of "Minimum" in public APIs to "Min" for consistency considering how many limits already use "Max".

So that means the MinimumDataRate class changes to MinDataRate, IHttpMinRequestBodyDataRateFeature.MinimumDataRate changes to IHttpMinRequestBodyDataRateFeature.MinDataRate, etc...

I don't care as much about private/internal names, but I think some of that could also change to match.

cesarbs replied:
I already change the property/param to bytesPerSecond, so that's taken care of. I'll change the type name now.

I went with Min because we already have Max on other properties.

halter73 commented:
I meant to leave a comment with my approval. I approve this pending changing "Minimum" to "Min" in all public APIs.

cesarbs commented:
Updated, will merge when CI is green.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -255,11 +255,11 @@ public TimeSpan RequestHeadersTimeout
         /// Defaults to 1 byte/second with a 5 second grace period.
         /// </remarks>
-        public MinimumDataRate RequestBodyMinimumDataRate { get; set; } = new MinimumDataRate(rate: 1, gracePeriod: TimeSpan.FromSeconds(5));
+        public MinDataRate MinRequestBodyDataRate { get; set; } = new MinDataRate(bytesPerSecond: 1, gracePeriod: TimeSpan.FromSeconds(5));
```
Change this to match the IIS default. I believe it's [240 bytes/second](https://www.iis.net/configreference/system.applicationhost/weblimits#005).

cesarbs replied:
Separate change. Let's discuss it in the office tomorrow.

halter73 replied:
And here I was trying to avoid creating a new issue 😢

halter73 replied:
https://github.com/aspnet/KestrelHttpServer/issues/1929

halter73 commented:
I know I already approved this. Hopeful changing the default is the last thing.
